### PR TITLE
Postgres migration

### PR DESCRIPTION
**Pull request by ice-myles**
Migration from tarantool to postgres database.

### CODE REVIEW

**Review by ankit-gautam23 — COMMENTED** [review]
LGTM

File: news/DDL.sql [hunk 1]
```diff
@@ -0,0 +1,42 @@
+-- SPDX-License-Identifier: ice License 1.0
+-- news
+CREATE TABLE IF NOT EXISTS news (
```
Comment (ice-ares) [inline news/DDL.sql:3]: Order of columns is: 

8 byte integers
4 byte integers
2 byte integers
Booleans
Arrays
Strings/or string derivates (like tsvector)

Also, nulls have to be last. If u have 2x 8 byte integers, 1 null & 1 non null, the non null one has to be first.


There's a sql query u can run to check if the column order in the ddl is optimal

File: news/DDL.sql [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                    created_at            TIMESTAMP NOT NULL,
+                    updated_at            TIMESTAMP NOT NULL,
+                    notification_channels TEXT,
```
Comment (ice-ares) [inline news/DDL.sql]: Cant remember if this is used or not

File: news/DDL.sql [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                    id                    TEXT NOT NULL,
+                    type                  TEXT NOT NULL,
+                    language              TEXT NOT NULL,
+                    title                 TEXT NOT NULL,
+                    image_url             TEXT NOT NULL,
+                    url                   TEXT NOT NULL UNIQUE,
+                    views                 BIGINT NOT NULL DEFAULT 0,
```
Comment (ice-ares) [inline news/DDL.sql]: Smallint

File: news/DDL.sql [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+                    PRIMARY KEY(language,id)
+                    );
+CREATE INDEX IF NOT EXISTS most_recent_news_lookup_ix ON news (language, type, created_at);
```
Comment (ice-ares) [inline news/DDL.sql]: Check if this is used, cuz i think in composite indexes the 2nd, 3rd, etc column can match only on =, not >/<

File: news/news_get.go [hunk 5]
```diff
@@ -4,36 +4,32 @@ package news
...
 						FROM news n
 							LEFT JOIN news_viewed_by_users nvu 
 								   ON nvu.language = n.language
 								  AND nvu.news_id = n.id
-								  AND nvu.user_id = :user_id
-						WHERE n.language = :language AND n.type = :type
+								  AND nvu.user_id = $1
+						WHERE n.language = $2 AND n.type = $3
 						ORDER BY nvu.created_at IS NULL DESC,
 								 n.created_at DESC
```
Comment (ice-ares) [inline news/news_get.go:28]: U can add desc to this column in the index ddl

File: news/news_get.go [hunk 6]
```diff
@@ -48,54 +44,43 @@ func (r *repository) GetUnreadNewsCount(ctx context.Context, language string, cr
...
 				   n.* 
 			FROM news n
 			      LEFT JOIN news_tags_per_news t
 			      		 ON t.language = n.language
 			      		AND t.news_id  = n.id
-			WHERE n.language = :language
-			  AND n.id = :id
+			WHERE n.language = $1
+			 		AND n.id = $2
+			GROUP BY n.created_at, n.updated_at, n.notification_channels, n.id, n.type, n.language, t.created_at
```
Comment (ice-ares) [inline news/news_get.go]: n.language, n.id is enough

File: news/news_modify.go [hunk 7]
```diff
@@ -131,32 +137,25 @@ func (r *repository) IncrementViews(ctx context.Context, newsID, language string
...
-	}
-	if err := r.sendNewsViewedMessage(ctx, tuple); err != nil {
-		bErr := errors.Wrapf(err, "failed to sendNewsViewedMessage for %#v", tuple)
-		ops = append(make([]tarantool.Op, 0, 1), tarantool.Op{Op: "-", Field: 9, Arg: uint64(1)}) //nolint:gomnd // That's the `views` column index.
-		if err = storage.CheckNoSQLDMLErr(r.db.UpdateTyped("NEWS", "pk_unnamed_NEWS_2", []any{language, newsID}, ops, &[]*News{})); err != nil {
-			return multierror.Append(bErr, errors.Wrapf(err, "[rollback]failed to decrement news views count for newsId:%v,language:%v", newsID, language)).ErrorOrNil() //nolint:lll,wrapcheck // .
-		}
-		if err = storage.CheckNoSQLDMLErr(r.db.DeleteTyped("NEWS_VIEWED_BY_USERS", "pk_unnamed_NEWS_VIEWED_BY_USERS_1", []any{language, tuple.NewsID, tuple.UserID}, &[]*ViewedNews{})); err != nil { //nolint:lll // .
-			return multierror.Append(bErr, errors.Wrapf(err, "[rollback]failed to delete NEWS_VIEWED_BY_USERS%#v", tuple)).ErrorOrNil() //nolint:wrapcheck // .
+	return errors.Wrapf(storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline news/news_modify.go]: Nope, too slow and this is gonna be called alot in the app, so follow the same approach as before

File: news/news_modify.go [hunk 8]
```diff
@@ -178,15 +177,14 @@ func (r *repository) sendNewsViewedMessage(ctx context.Context, vn *ViewedNews)
...
-		switch tErr.Data[storage.IndexName] {
-		case "unique_unnamed_NEWS_1":
+		if storage.IsErr(err, storage.ErrDuplicate, "url") { //nolint:gocritic // Switch case not possible.
 			field = "url"
-		case "pk_unnamed_NEWS_2":
-			field = "language"
-		default:
-			log.Panic("unexpected indexName `%v` for news space", tErr.Data[storage.IndexName])
+		} else if storage.IsErr(err, storage.ErrDuplicate, "pk") {
+			field = "pk"
```
Comment (ice-ares) [inline news/news_modify.go]: id

File: notifications/DDL.sql [hunk 9]
```diff
@@ -0,0 +1,51 @@
+-- SPDX-License-Identifier: ice License 1.0
+--************************************************************************************************************************************
+-- users
+CREATE TABLE IF NOT EXISTS users  (
+                    last_ping_cooldown_ended_at             TIMESTAMP,
+                    disabled_push_notification_domains      TEXT,
```
Comment (ice-ares) [inline notifications/DDL.sql]: We can use []text

File: notifications/DDL.sql [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+                    phone_number                            TEXT,
+                    email                                   TEXT,
+                    first_name                              TEXT,
+                    last_name                               TEXT,
+                    user_id                                 TEXT NOT NULL primary key,
+                    username                                TEXT,
+                    profile_picture_name                    TEXT,
+                    referred_by                             TEXT,
+                    phone_number_hash                       TEXT,
+                    agenda_phone_number_hashes              TEXT,
```
Comment (ice-ares) [inline notifications/DDL.sql]: This is changed in a different pr right?

File: notifications/email_notifications.go [hunk 11]
```diff
@@ -31,23 +29,19 @@ func (r *repository) sendEmailNotification(ctx context.Context, en *emailNotific
...
-		en.en.From.Name = internationalizedEmailDisplayNames["en"]
-	}
-	if err := r.emailClient.Send(ctx, en.en, email.Participant{Name: en.displayName, Email: en.sn.NotificationChannelValue}); err != nil {
-		return multierror.Append( //nolint:wrapcheck // Not needed.
-			errors.Wrapf(err, "failed to send email notification:%#v, desired to be sent:%#v", en.en, en.sn),
-			errors.Wrapf(storage.CheckNoSQLDMLErr(r.db.DeleteTyped("SENT_NOTIFICATIONS", "pk_unnamed_SENT_NOTIFICATIONS_1", &en.sn.sentNotificationPK, &[]*sentNotification{})), //nolint:lll // .
-				"failed to delete SENT_NOTIFICATIONS as a rollback for %#v", en.sn),
-		).ErrorOrNil()
-	}
+	return errors.Wrapf(storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline notifications/email_notifications.go]: Nope, follow the previous pattern.

Transactions are very slow, we use them very rarely and only in places where we dont have any other choice or places not used that much

File: notifications/inapp_notifications.go [hunk 12]
```diff
@@ -23,42 +22,34 @@ type (
...
-	}
 
-	if err := r.personalInAppFeed.Send(ctx, in.in, in.sn.UserID); err != nil {
-		return multierror.Append( //nolint:wrapcheck // Not needed.
-			errors.Wrapf(err, "failed to send inApp notification:%#v, desired to be sent:%#v", in.in, in.sn),
-			errors.Wrapf(storage.CheckNoSQLDMLErr(r.db.DeleteTyped("SENT_NOTIFICATIONS", "pk_unnamed_SENT_NOTIFICATIONS_1", &in.sn.sentNotificationPK, &[]*sentNotification{})), //nolint:lll // .
-				"failed to delete SENT_NOTIFICATIONS as a rollback for %#v", in.sn),
-		).ErrorOrNil()
-	}
+	return errors.Wrapf(storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline notifications/inapp_notifications.go]: Same thing here

File: notifications/push_notifications.go [hunk 13]
```diff
@@ -147,27 +139,27 @@ func (r *repository) sendPushNotification(ctx context.Context, pn *pushNotificat
...
-		return errors.Wrapf(err, "failed to insert %#v", pn.sn)
-	}
-	responder := make(chan error, 1)
-	defer close(responder)
-	r.pushNotificationsClient.Send(ctx, pn.pn, responder)
-	if err := <-responder; err != nil {
-		var cErr error
-		if errors.Is(err, push.ErrInvalidDeviceToken) {
-			cErr = r.clearInvalidPushNotificationToken(ctx, pn.sn.UserID, pn.pn.Target)
+	err := storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline notifications/push_notifications.go]: Same here

File: notifications/push_notifications.go [hunk 14]
```diff
@@ -210,16 +197,12 @@ func (r *repository) broadcastPushNotification(ctx context.Context, bpn *broadca
...
-	if err := r.pushNotificationsClient.Broadcast(ctx, bpn.pn); err != nil {
-		return multierror.Append( //nolint:wrapcheck // Not needed.
-			errors.Wrapf(err, "failed to broadcast push notification:%#v, desired to be sent:%#v", bpn.pn, bpn.sa),
-			errors.Wrapf(storage.CheckNoSQLDMLErr(r.db.DeleteTyped("SENT_ANNOUNCEMENTS", "pk_unnamed_SENT_ANNOUNCEMENTS_1", &bpn.sa.sentAnnouncementPK, &[]*sentAnnouncement{})), //nolint:lll // .
-				"failed to delete SENT_ANNOUNCEMENTS as a rollback for %#v", bpn.sa),
-		).ErrorOrNil()
-	}
 
-	return nil
+	return errors.Wrapf(storage.DoInTransaction(ctx, r.db, func(conn storage.QueryExecer) error {
```
Comment (ice-ares) [inline notifications/push_notifications.go]: Same here

File: notifications/users.go [hunk 15]
```diff
@@ -194,33 +195,30 @@ func (r *repository) ToggleNotificationChannelDomain( //nolint:funlen,gocognit,g
...
 		}
 	}
-	op.Arg = &sanitizedDisabledDomains
-	resp := make([]*user, 0, 1)
-	if err = storage.CheckNoSQLDMLErr(r.db.UpdateTyped("USERS", "pk_unnamed_USERS_1", tarantool.StringKey{S: userID}, append(make([]tarantool.Op, 0, 1), op), &resp)); err != nil { //nolint:lll // .
-		if errors.Is(err, storage.ErrNotFound) {
-			err = ErrRelationNotFound
+	valuesForUpdate = &sanitizedDisabledDomains
+	sql := fmt.Sprintf(`UPDATE users SET %v where user_id = $1`, fieldForUpdate)
+	if rowsUpdated, tErr := storage.Exec(ctx, r.db, sql, append([]any{userID}, valuesForUpdate)...); tErr != nil {
```
Comment (ice-ares) [inline notifications/users.go]: || rowsUpdated ==0

File: notifications/users.go [hunk 16]
```diff
@@ -194,33 +195,30 @@ func (r *repository) ToggleNotificationChannelDomain( //nolint:funlen,gocognit,g
...
+		if rowsUpdated == 0 || storage.IsErr(tErr, storage.ErrNotFound) {
```
Comment (ice-ares) [inline notifications/users.go]: err==nil or ...

File: notifications/users.go [hunk 17]
```diff
@@ -364,36 +370,39 @@ func (r *repository) PingUser(ctx context.Context, userID string) error { //noli
...
 
 	if err = r.sendUserPingMessage(ctx, up); err != nil {
-		params["now_nanos"] = usr.LastPingCooldownEndedAt
-		params["last_ping_cooldown_ended_at"] = newPingCooldownEndsAt
-		rErr := storage.CheckSQLDMLErr(r.db.PrepareExecute(sql, params))
-		if rErr != nil && errors.Is(rErr, storage.ErrNotFound) {
+		params[0] = usr.LastPingCooldownEndedAt
+		params[3] = newPingCooldownEndsAt
+		rRowsUpdated, rErr := storage.Exec(ctx, r.db, sql, params...)
+		if rRowsUpdated == 0 || (rErr != nil && storage.IsErr(rErr, storage.ErrNotFound)) {
```
Comment (ice-ares) [inline notifications/users.go]: err == nil and rowsupdsted ==0

File: notifications/users.go [hunk 18]
```diff
@@ -364,36 +370,39 @@ func (r *repository) PingUser(ctx context.Context, userID string) error { //noli
...
+		    WHERE user_id = $2
+			  AND referred_by = $3
+			  AND COALESCE(last_ping_cooldown_ended_at, to_timestamp(0)) = COALESCE($4, to_timestamp(0))`
+	params := []any{
+		newPingCooldownEndsAt.Time,
+		userID,
+		usr.ReferredBy,
+		lastPingedTime,
+	}
+	if rowsUpdated, sErr := storage.Exec(ctx, r.db, sql, params...); rowsUpdated == 0 || (sErr != nil && storage.IsErr(sErr, storage.ErrNotFound)) {
```
Comment (ice-ares) [inline notifications/users.go]: err == nil and rowsupdated ==0

**Review by ice-ares — CHANGES_REQUESTED** [review]

File: notifications/DDL.sql (hunk 9 above)
Comment (ice-cronus) [reply notifications/DDL.sql]: It is users.Enum initially from eskimo. And we're using it for the HiddenProfileElements on eskimo (which is TEXT and parsed on Scan() fuction), and in santa as well. But as for me we should follow the same pattern for datatype everywhere, so, we're keeping TEXT here or convert that HiddenProfileElements and santa's enums (achieved badges, levels, etc) to array as well, probably the second approach is better.

Comment (ice-ares) [reply notifications/DDL.sql]: I’d say make everything that is comma separated into []text, across all services

File: news/DDL.sql (hunk 4 above)
Comment (ice-myles) [reply news/DDL.sql]: This index is used.

File: notifications/DDL.sql (hunk 10 above)
Comment (ice-myles) [reply notifications/DDL.sql]: Right

File: news/DDL.sql (hunk 2 above)
Comment (ice-myles) [reply news/DDL.sql]: Is used at the notifications processor.

File: news/news_modify.go [hunk 19]
```diff
@@ -65,38 +65,45 @@ func (r *repository) updateNews(ctx context.Context, news *TaggedNews) error { /
...
+	sql += fmt.Sprintf(" WHERE ID = $%v AND LANGUAGE = $%v", fieldIndex, fieldIndex+1)
+	fieldIndex += 2
 	if lu := lastUpdatedAt(ctx); lu != nil {
-		params["lastUpdatedAt"] = lu
-		sql += " AND UPDATED_AT = :lastUpdatedAt"
+		args = append(args, lu.Time)
+		sql += fmt.Sprintf(" AND UPDATED_AT = $%v", fieldIndex)
 	}
-	if err := storage.CheckSQLDMLErr(r.db.PrepareExecute(sql, params)); err != nil {
-		if err = detectAndParseDuplicateDatabaseError(err); errors.Is(err, storage.ErrNotFound) {
```
Comment (ice-ares) [inline news/news_modify.go:95]: What about this?

File: news/news_modify.go [hunk 20]
```diff
@@ -131,26 +138,28 @@ func (r *repository) IncrementViews(ctx context.Context, newsID, language string
...
 		}
 
 		return errors.Wrapf(err, "failed to increment news views count for newsId:%v,language:%v", newsID, language)
 	}
 	if err := r.sendNewsViewedMessage(ctx, tuple); err != nil {
 		bErr := errors.Wrapf(err, "failed to sendNewsViewedMessage for %#v", tuple)
-		ops = append(make([]tarantool.Op, 0, 1), tarantool.Op{Op: "-", Field: 9, Arg: uint64(1)}) //nolint:gomnd // That's the `views` column index.
-		if err = storage.CheckNoSQLDMLErr(r.db.UpdateTyped("NEWS", "pk_unnamed_NEWS_2", []any{language, newsID}, ops, &[]*News{})); err != nil {
-			return multierror.Append(bErr, errors.Wrapf(err, "[rollback]failed to decrement news views count for newsId:%v,language:%v", newsID, language)).ErrorOrNil() //nolint:lll,wrapcheck // .
+		sql = `UPDATE news SET views = views - 1 WHERE language = $1 AND id = $2`
```
Comment (ice-ares) [inline news/news_modify.go]: Make sure this doesnt become nevative

File: notifications/inapp_notifications.go [hunk 21]
```diff
@@ -23,54 +22,44 @@ type (
...
 
 	return nil
 }
 
-func (r *repository) broadcastInAppNotification(ctx context.Context, bin *broadcastInAppNotification) error { //nolint:dupl // Wrong.
+func (r *repository) broadcastInAppNotification(ctx context.Context, bin *broadcastInAppNotification) error {
 	if ctx.Err() != nil {
 		return errors.Wrap(ctx.Err(), "unexpected deadline")
 	}
-	if true {
```
Comment (ice-ares) [inline notifications/inapp_notifications.go:52]: Keep this and i believe there's something similar in an email flow somewjere, keep that as well

File: notifications/notifications.go [hunk 22]
```diff
@@ -48,21 +48,14 @@ func StartProcessor(ctx context.Context, cancel context.CancelFunc) Processor {
...
+		db:                      storage.MustConnect(context.Background(), ddl, applicationYamlKey), //nolint:contextcheck // We need to gracefully shut it down.
 		mb:                      messagebroker.MustConnect(ctx, applicationYamlKey),
 		pushNotificationsClient: push.New(applicationYamlKey),
 		pictureClient:           picture.New(applicationYamlKey),
 		emailClient:             email.New(applicationYamlKey),
-		/*
-			personalInAppFeed:       inapp.New(applicationYamlKey, "notifications"),
-			globalInAppFeed:         inapp.New(applicationYamlKey, "announcements"),
-		*/
+		personalInAppFeed:       inapp.New(applicationYamlKey, "notifications"),
```
Comment (ice-ares) [inline notifications/notifications.go]: Revert

File: notifications/notifications.go [hunk 23]
```diff
@@ -244,3 +233,81 @@ func executeConcurrently(fs ...func() error) error {
...
+		sn.NotificationType,
+		sn.NotificationChannel,
+		sn.NotificationChannelValue,
+	)
+
+	return errors.Wrapf(err, "failed to insert sent notification %#v", sn)
+}
+
+func (r *repository) insertSentAnnouncement(ctx context.Context, sa *sentAnnouncement) error {
+	sql := `INSERT INTO sent_announcements (SENT_AT, LANGUAGE, UNIQUENESS, NOTIFICATION_TYPE, NOTIFICATION_CHANNEL, NOTIFICATION_CHANNEL_VALUE)
```
Comment (ice-ares) [inline notifications/notifications.go]: Be consistent, use the same formatting everywhere

**Review by ice-ares — CHANGES_REQUESTED** [review]

File: notifications/inapp_notifications.go [hunk 24]
```diff
@@ -23,45 +22,45 @@ type (
...
 		return errors.Wrapf(err, "failed to insert %#v", in.sn)
 	}
 
 	if err := r.personalInAppFeed.Send(ctx, in.in, in.sn.UserID); err != nil {
-		return multierror.Append( //nolint:wrapcheck // Not needed.
+		return multierror.Append(
 			errors.Wrapf(err, "failed to send inApp notification:%#v, desired to be sent:%#v", in.in, in.sn),
-			errors.Wrapf(storage.CheckNoSQLDMLErr(r.db.DeleteTyped("SENT_NOTIFICATIONS", "pk_unnamed_SENT_NOTIFICATIONS_1", &in.sn.sentNotificationPK, &[]*sentNotification{})), //nolint:lll // .
-				"failed to delete SENT_NOTIFICATIONS as a rollback for %#v", in.sn),
-		).ErrorOrNil()
```
Comment (ice-ares) [inline notifications/inapp_notifications.go:42]: u need this

File: notifications/inapp_notifications.go [hunk 25]
```diff
@@ -23,45 +22,45 @@ type (
...
 		return errors.Wrapf(err, "failed to insert %#v", bin.sa)
 	}
 
 	if err := r.globalInAppFeed.Send(ctx, bin.in, bin.sa.NotificationChannelValue); err != nil {
-		return multierror.Append( //nolint:wrapcheck // Not needed.
+		return multierror.Append(
 			errors.Wrapf(err, "failed to broadcast inApp notification:%#v, desired to be sent:%#v", bin.in, bin.sa),
-			errors.Wrapf(storage.CheckNoSQLDMLErr(r.db.DeleteTyped("SENT_ANNOUNCEMENTS", "pk_unnamed_SENT_ANNOUNCEMENTS_1", &bin.sa.sentAnnouncementPK, &[]*sentAnnouncement{})), //nolint:lll // .
-				"failed to delete SENT_ANNOUNCEMENTS as a rollback for %#v", bin.sa),
-		).ErrorOrNil()
```
Comment (ice-ares) [inline notifications/inapp_notifications.go:64]: u need this

**Pull request merged**